### PR TITLE
TICKET-T-16950-2: Fix Location permission issue

### PR DESCRIPTION
**Pull request by mohan-manthri**
- LocationPermission will be requested only once per app launch if not granted already.
- Code improvements.

### CODE REVIEW

File: lib/positioning/positioning_engine.dart [hunk 1]
```diff
@@ -119,20 +144,20 @@ class PositioningEngine {
...
-    } else {
+    final bool didLocationServicesEnabled = await _didLocationServicesEnabled;
+    if (didLocationServicesEnabled && _locationEngine != null) {
+      // As location engine is already created, we do not need to create
+      // a new one.
+      return;
+    }
+    if (!didLocationServicesEnabled) {
       _locationEngineStatusController.add(LocationEngineStatus.notAllowed);
+    } else if (didLocationServicesEnabled && _locationEngine == null) {
```
Comment (Guru-cm-here) [inline lib/positioning/positioning_engine.dart:159]: We can make a few improvements to enhance readability and remove unnecessary conditions.

```
Future<void> _checkLocationServicesStatus() async {
  final bool didLocationServicesEnabled = await _didLocationServicesEnabled;
  
  if (didLocationServicesEnabled && _locationEngine != null) {
    return; // As the location engine is already created, we do not need to create a new one.
  }

  if (!didLocationServicesEnabled) {
    _locationEngineStatusController.add(LocationEngineStatus.notAllowed);
  } else {
    await _createLocationEngineIfPermissionsGranted();
  }
}
```

File: lib/positioning/positioning_engine.dart [hunk 2]
```diff
@@ -102,6 +108,25 @@ class PositioningEngine {
     }
   }
 
+  /// Returns [true] if both [Permission.location] and [Permission.locationAlways]
+  /// are granted, otherwise returns [false].
+  ///
+  /// Returns [false] if location services is not enabled.
+  Future<bool> _didLocationPermissionsGranted() async {
```
Comment (Guru-cm-here) [inline lib/positioning/positioning_engine.dart:115]: Improvements to enhance readability and remove unnecessary await.
Combined the two separate if statements into one

Please feel free to adopt the suggestions according to the requirements.

```
Future<bool> _didLocationPermissionsGranted() async {
  final bool isLocationServicesEnabled = await _didLocationServicesEnabled;
  if (!isLocationServicesEnabled) {
    return false;
  }

  final bool isLocationPermissionGranted = await Permission.location.isGranted;
  if (Platform.isAndroid && await getAndroidApiVersion() >= _androidApiLevel30) {
    final bool isLocationAlwaysPermissionGranted = await Permission.locationAlways.status.isGranted;
    return isLocationPermissionGranted && isLocationAlwaysPermissionGranted;
  }
  return isLocationPermissionGranted;
}
```

File: lib/positioning/positioning_engine.dart (hunk 1 above)
Comment (Guru-cm-here) [reply lib/positioning/positioning_engine.dart:159]: Please feel free to adopt the suggestions according to the requirements.

**Review by mr-pant — APPROVED** [review]

File: lib/positioning/positioning_engine.dart (hunk 1 above)
Comment (Guru-cm-here) [reply lib/positioning/positioning_engine.dart:159]: We can have another PR for code improvement

**Review by Guru-cm-here — APPROVED** [review]

**Pull request merged**